### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by franclobo**
## Processing data in models
- Create models
- Use models to insert data
- Create custom methods

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,5 @@
 class User < ApplicationRecord
+  has_many :comments
+  has_many :likes
+  has_many :posts
```
Comment (rloterh) [inline app/models/user.rb]: - [x] As per the given _ERD_, the _User_ **association** should have _author_id_ referenced as **foreign key** in _Post_, _Comment,_ and _Like_ models. Kindly ensure to improve this aspect, similar to the code example below:
```ruby
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/post.rb [hunk 2]
```diff
@@ -1,2 +1,5 @@
 class Post < ApplicationRecord
+  belongs_to :author, class_name: :User, foreign_key: :user_id
```
Comment (rloterh) [inline app/models/post.rb]: - [x] Foreign key isn't required here, as it should be already established in the `user` model. Simply, _Post_ belongs to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/comment.rb [hunk 3]
```diff
@@ -1,2 +1,4 @@
 class Comment < ApplicationRecord
+  belongs_to :author, class_name: :User, foreign_key: :user_id
```
Comment (rloterh) [inline app/models/comment.rb]: - [x] Foreign key isn't required here, as it should be already established in the `user` model. Simply, _Comments_ belongs to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/comment.rb [hunk 4]
```diff
@@ -1,2 +1,4 @@
...
+  belongs_to :post, class_name: :Post, foreign_key: :post_id
```
Comment (rloterh) [inline app/models/comment.rb]: - [x] Foreign key isn't required at this point either. Simply do:
```ruby
belongs_to :post

File: app/models/like.rb [hunk 5]
```diff
@@ -1,2 +1,4 @@
 class Like < ApplicationRecord
+  belongs_to :users
```
Comment (rloterh) [inline app/models/like.rb]: - [x] As per the given _ERD_, the _Like_ **association** should have _Like_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/user.rb [hunk 6]
```diff
@@ -1,2 +1,5 @@
 class User < ApplicationRecord
```
Comment (rloterh) [inline app/models/user.rb:1]: - [x] This module requires that `User` model has method that returns 3 most recent posts. This should be similar to the example below:

```ruby
# app/models/user.rb
...

def recent_post
  posts.order('created_at Desc').limit(3)
end
```

File: app/models/post.rb [hunk 7]
```diff
@@ -1,2 +1,5 @@
 class Post < ApplicationRecord
```
Comment (rloterh) [inline app/models/post.rb:1]: - [x] This module requires that `Post` model has method that updates the posts counter for a user. This should be similar to the example below:

```ruby
# app/models/post.rb
...

def update_post_counter
  author.increment!(:posts_counter)
end
```

Comment (rloterh) [inline app/models/post.rb:1]: - [x] `Post` model should have method that returns the 5 most recent comments for a given post. This should be similar to the example below:

```ruby
# app/models/post.rb
...

def recent_comments
  comments.order('created_at Desc').limit(5)
end
```

File: app/models/comment.rb [hunk 8]
```diff
@@ -1,2 +1,4 @@
 class Comment < ApplicationRecord
```
Comment (rloterh) [inline app/models/comment.rb:1]: - [ ] `Comment` model should have method that updates comments counter for a post. This should be similar to the example below:

```ruby
# app/models/comment.rb
...

def update_comments_counter
  post.increment!(:comments_counter)
end
```

File: app/models/like.rb [hunk 9]
```diff
@@ -1,2 +1,4 @@
 class Like < ApplicationRecord
```
Comment (rloterh) [inline app/models/like.rb:1]: - [x] `Like` model should have method that updates the likes counter for a post. This should be similar to the example below:

```ruby
# app/models/like.rb
...

def update_likes_counter
  post.increment!(:likes_counter)
end
```

Comment (rloterh) [inline app/models/like.rb:1]: - [ ] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_likes_counter method every time a new like has been saved to database, then make the `update_likes_counter` method private, similar to:

```ruby
 # app/models/likes.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

File: app/models/comment.rb (hunk 8 above)
Comment (rloterh) [inline app/models/comment.rb:1]: - [x] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_comments_counter method every time a new like has been saved to database, then make the `update_comments_counter` method private, similar to:

```ruby
 # app/models/comments.rb
  after_save :update_comments_counter

  private

  def update_comments_counter
    author.increment!(:comments_counter)
  end
```

File: app/models/post.rb (hunk 7 above)
Comment (rloterh) [inline app/models/post.rb:1]: - [x] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_posts_counter method every time a new like has been saved to database, then make the `update_posts_counter` method private, similar to:

```ruby
 # app/models/posts.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Francisco, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- No linter errors


## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Kindly check the comments marked **[OPTIONAL]** under the review._


<br>
Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 1 above)
Comment (franclobo) [reply app/models/user.rb]: Done

File: app/models/post.rb (hunk 2 above)
Comment (franclobo) [reply app/models/post.rb]: Done

File: app/models/comment.rb (hunk 3 above)
Comment (franclobo) [reply app/models/comment.rb]: Done

File: app/models/comment.rb (hunk 4 above)
Comment (franclobo) [reply app/models/comment.rb]: Done

File: app/models/like.rb (hunk 5 above)
Comment (franclobo) [reply app/models/like.rb]: Done

File: app/models/user.rb (hunk 6 above)
Comment (franclobo) [reply app/models/user.rb:1]: Done

File: app/models/post.rb (hunk 7 above)
Comment (franclobo) [reply app/models/post.rb:1]: Done

Comment (franclobo) [reply app/models/post.rb:1]: Done

File: app/models/comment.rb (hunk 8 above)
Comment (franclobo) [reply app/models/comment.rb:1]: Done

File: app/models/like.rb (hunk 9 above)
Comment (franclobo) [reply app/models/like.rb:1]: Done

Comment (franclobo) [reply app/models/like.rb:1]: Done

File: app/models/comment.rb (hunk 8 above)
Comment (franclobo) [reply app/models/comment.rb:1]: Done

File: app/models/post.rb (hunk 7 above)
Comment (franclobo) [reply app/models/post.rb:1]: Done

**Review by rotimiazeez — APPROVED** [review]
# Hi @franclobo 👍🏽 , 

### Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**